README
Bio::Graphics - Generate GD images of Bio::Seq objects

SYNOPSIS

This is a simple GD-based renderer (diagram drawer) for DNA and
protein sequences.  It is used from Perl like this:

  use Bio::Graphics;
  use Bio::DB::BioFetch;  # or some other Bio::SeqI generator
  # get a Bio::SeqI object somehow
  my $bf     = Bio::DB::BioFetch->new;
  my $cosmid = $bf->get_Seq_by_id('CEF58D5');

  my @features = $cosmid->all_SeqFeatures;
  my @CDS      = grep {$_->primary_tag eq 'CDS'}  @features;
  my @gene     = grep {$_->primary_tag eq 'gene'} @features;
  my @tRNAs    = grep {$_->primary_tag eq 'tRNA'} @features;
  # let the drawing begin...
  my $panel = Bio::Graphics::Panel->new(
				      -segment => $cosmid,
				      -width  => 800
				     );

  $panel->add_track(arrow => $cosmid,
	  	   -bump => 0,
		   -double=>1,
		   -tick => 2);

  $panel->add_track(transcript  => \@gene,
		   -bgcolor    =>  'blue',
		   -fgcolor    =>  'black',
		   -key        => 'Genes',
		   -bump       =>  +1,
		   -height     =>  10,
		   -label      => 1,
		   -description=> 1
		 ) ;

  $panel->add_track(transcript2  => \@CDS,
		    -bgcolor    =>  'cyan',
		    -fgcolor    =>  'black',
		    -key        => 'CDS',
		    -bump       =>  +1,
		    -height     =>  10,
		    -label      => 1,
		    -description=> 1,
		 );

  $panel->add_track(generic    => \@tRNAs,
		    -bgcolor   =>  'red',
		    -fgcolor   =>  'black',
		    -key       => 'tRNAs',
		    -bump      =>  +1,
		    -height    =>  8,
		    -label      => 1,
		   );

  my $gd = $panel->gd;
  print $gd->can('png') ? $gd->png : $gd->gif;

DESCRIPTION

The Bio::Graphics::Panel class provides drawing and formatting
services for any object that implements the Bio::SeqFeatureI
interface, including Ace::Sequence::Feature, Das::Segment::Feature and
Bio::DB::Graphics objects.  It can be used to draw sequence
annotations, physical (contig) maps, protein domains, or any other
type of map in which a set of discrete ranges need to be laid out on
the number line.

The module supports a drawing style in which each type of feature
occupies a discrete "track" that spans the width of the display.  Each
track will have its own distinctive "glyph", a configurable graphical
representation of the feature.

The module also supports a more flexible style in which several
different feature types and their associated glyphs can occupy the
same track.  The choice of glyph is under run-time control.

Semantic zooming (for instance, changing the type of glyph depending
on the density of features) is supported by a callback system for
configuration variables.  The module has built-in support for Bio::Das
stylesheets, and stylesheet-driven configuration can be intermixed
with semantic zooming, if desired.

You can add a key to the generated image using either of two key
styles.  One style places the key captions at the top of each track.
The other style generates a graphical key at the bottom of the image.

Note that this modules depends on Bio::Perl, GD and the
Text::Parsewords module.

A good tutorial on this module can be found at
http://www.bioperl.org/wiki/HOWTO:Graphics.

BUILDING AND INSTALLING

From the directory in which this README file is located:

  % perl ./Build.PL
  % ./Build test
  % ./Build install

The last step may need to be run as root.

You can test whether the module is working by running the testit.pl
script located in the eg subdirectory.  It produces a PNG file, so you
should pipe the output to an image displayer, such as xv:

  eg/testit.pl | xv -

THE FEATURE_DRAW.PL SCRIPT

This module contains a simple script named "feature_draw.pl" that
takes a file of annotations and generates an image.  When first built,
the script lives in the "scripts" subdirectory.  After installation,
the script will be moved into your system-wide scripts directory
(usually /usr/bin or something similar).

There is a sample features file in the "eg" subdirectory.  After
building, but before installation, you can run the following command
from the directory in which this README file is located:

  %  scripts/feature_draw.pl eg/feature_data.txt | display -

This is piping the output directly to the "display" program, part of
the ImageMagick package.  Replace with your favorite PNG-capable
display program.

After installation, feature_draw.pl will be available on your command
path.  Run "man feature_draw.pl" to see the manual page, or use the -h
option to get a quick summary of usage.

THE GENERIC GENOME BROWSER

The gbrowse script, part of the Generic Genome Browser package, has
been moved into its own package.  It creates a high-performance
web-based genome browser based on the Bio::Graphics package.  You can
get it by anonymous CVS to SourceForge:

 cvs -d:pserver:anonymous@cvs.gmod.sourceforge.net:/cvsroot/gmod \
	co Generic-Genome-Browser

Lincoln Stein <lincoln.stein@oicr.on.ca>
12 December 2008


Changes
Revision history for Perl extension Bio::Graphics.
2.39
   - Turn off image-based regression tests becuase libgd has started to produce visually-identical
     images that are different at the byte level.
	
2.38
   - If left labeling requested and the selected font size is larger than the available height,
     then go to smallest available font (gdTinyFont).

2.37
   - Merge fix for incorrect parsing of feature_file sections (issue #6)
   - Added decorated_transcript glyph
   - Added nathanweeks fix for broken heatmap glyph

2.36
   - Fix regression in glyph parent_feature() method so that it once again returns
     the parent of the current feature.
   - Remove recommended prerequisite of Bio::SCF, which is hardly used now.

2.35
   - Change Glyph/segments.pm to work with both variants of CIGAR strings.	
   - Workaround for broken useFontConfig() support in versions of GD prior to 2.50.
   - Fix "gdTinyFont doesn't support height() method error in xyplot".

2.34 Thu May 16 15:42:25 CDT 2013
    - Fixed silent crashes when rendering with the GD::SVG class.

2.33 Fri Feb 22 15:58:10 EST 2013
    - Add truetype support. Enable by passing -truetype=>1 to Bio::Graphics::Panel->new()

2.32 Mon Dec 10 05:47:45 EST 2012
   - Clean up appearance of crossbox to avoid odd black bar in the middle.

2.32 Mon Dec 10 05:47:45 EST 2012
   - Clean up appearance of crossbox to avoid odd black bar in the middle.

2.31 Tue Sep 25 22:39:43 EDT 2012
   - Fix infinite loop when drawing wiggle_xyplots with z-score scaling
     over regions that do not vary much.

2.30
    - Added the glyph for FlyBase's "stacked wiggle" plot (fb_shmiggle.pm) as well as a utility script for preparing the data (index_cov_files.pl).

2.29
    - Fixes for Bio::Graphics::Wiggle to prevent crash when the "bottom" keystyle selected.
    - Add ability to obtain Bio::Graphics::Wiggle features directly from the 
    Bio::Graphics::Wiggle::Loader without creating an intermediate gff3 file.

2.28
    - Fixes to restore density plot functionality in hybrid_plot and vista_plot.
    - Fix handling of Bio::Graphics::Wiggle files in xyplots.

2.27
    - Fix handling of code ref subs to allow for option=>\&Packagename::functionname.

2.26
    - Support for normalizing quantitative plots across entire track.
    - Support for transparency within tracks, allowing features to overlap.
    - Support for -color_series and -color_cycle options (see Bio::Graphics::Panel)
      which cycle dynamically among a fixed series of background colors.

2.25
    - Deprecate xyplot "histogram" subtype and default to "boxes".
    - Add xyplot overlay feature.
    - Fix gene glyph so that it collapses appropriately when bump explicitly
      set to 0.
    - Highlighting now works on features contained within groups.

2.24
    - In feature file configuration sections, allow a tag to contain ":".
  
2.23
    - Workaround a problem in BioPerl 1.069 which prevents FeatureFile from rendering
      properly. Fixed in 1.06902 and higher.
    - Fix issue which caused the track bgcolor property to be ignored in some cases.

2.22
     - Changed package "Math" to "Bio::Graphics::Math" to avoid CPAN namespace
       collisions.

2.21
     - Changed almost all occurrences of attributes() into eval{$feature->get_tag_values()}
       to achieve compatibility with Bio::SeqFeatureI. One exception is in spectrogram glyph,
       which depends on the non-standard behavior of attributes() when called with no args.
 
2.20
     - Fixed the various "wiggle" glyphs so that scale is always drawn even if no data is to be
       shown in selected region.

2.19
     - Made "record_label_positions" default to "true" for wiggle_density plot. This activates
       a fixed-position floating label when density plots are being shown in GBrowse.
     - Fixed a bug in wiggle_xyplot that prevented plot from being drawn in SVG renderings.

2.18
     - Made handling of min/max scale calculations consistent across all xyplot glyphs and their derivatives.
     - Added following autoscale options to wiggle_xyplot and wiggle_whiskers: "z-score" to rescale data such
       that mean is zero and values are standard deviation-fold change; and "clipped_global" to scale to
       global mean +/- some number of standard deviations indicated by "z_score_bounds".
     - Added "z_score_bounds" option to wiggle_xyplot and wiggle_whiskers to control how many standard
       deviations to show.
       

2.17
     - In segments glyph, fixed bleedover of "deletion" color when a deletion is followed by an insertion.
     - In segments glyph, fixed the display of inserted bases such that they will not bleed into a
       preceding deletion.
     - In segments glyph, fixed the mismatch color highlighting when the alignment is to the negative
       strand.
     - In segments glyph, fixed the default colors for mismatch, insertion and deletion.
     - In segments glyph, fixed treatment of soft clipping as an "insertion"; this will avoid the insertion
       color from appearing at the ends of soft-clipped reads.
     - In gene glyph, fixed occasional linking of two neighboring transcripts.
     
2.16
     - Distinguish between "chromosome" and "global" autoscaling in the xyplot and density glyphs.
       Global autoscaling only works when underlying database is bigwig.
     - Regularized options which select glyph subtypes with an option named "glyph_subtype".

2.15
     - Fixed documentation bug: sort_order options to sort by feature length should be
       "longest" and "shortest" rather than "longer and "shorter".
     - Improved layout algorithm, achieving speedup of ~4x on busy tracks.
     - Fixed font color problems when displaying multiple alignments at DNA level in
       the segments glyph.
     - Fixed problem of DNA alignment indels disappearing from view when they span entire region.
     - Fixed problem of mismatch and indel colors leaking off ends of feature arrows.

2.14

eg/testit.pl
#!/usr/bin/perl -w

use lib '.','..','./lib','./blib/lib','../blib/lib';
use strict;

use Bio::Graphics::Panel;
use Bio::Graphics::Feature;

my $ftr = 'Bio::Graphics::Feature';

my $segment = $ftr->new(-start=>-100,-end=>1000,-name=>'ZK154',-type=>'clone');
my $zk154_1 = $ftr->new(-start=>-50,-end=>800,-name=>'ZK154.1',-type=>'gene');
my $zk154_2 = $ftr->new(-start=>380,-end=>500,-name=>'ZK154.2',-type=>'gene');
my $zk154_3 = $ftr->new(-start=>900,-end=>1200,-name=>'ZK154.3',-type=>'gene');

my $zed_27 = $ftr->new(-segments=>[[400,500],[550,600],[800,950]],
		   -name=>'zed-27',
		   -subtype=>'exon',-type=>'transcript');
my $abc3 = $ftr->new(-segments=>[[100,200],[350,400],[500,550]],
		    -name=>'abc53',
		     -strand => -1,
		    -subtype=>'exon',-type=>'transcript');
my $xyz4 = $ftr->new(-segments=>[[40,80],[100,120],[200,280],[300,320]],
		     -name=>'xyz4',
		     -subtype=>'predicted',-type=>'alignment');

my $m3 = $ftr->new(-segments=>[[20,40],[30,60],[90,270],[290,300]],
		   -name=>'M3',
		   -subtype=>'predicted',-type=>'alignment');

my $bigone = $ftr->new(-segments=>[[-200,-120],[90,270],[290,300]],
		   -name=>'big one',
		   -subtype=>'predicted',-type=>'alignment');

my $fred_12 = $ftr->new(-segments=>[$xyz4,$zed_27],
			-type => 'group',
			-name =>'fred-12');

my $confirmed_exon1 = $ftr->new(-start=>1,-stop=>20,
				-type=>'exon',
				-source=>'confirmed',
				-name => 'confirmed1',
			       );
my $predicted_exon1 = $ftr->new(-start=>30,-stop=>50,
				-type=>'exon',
				-name=>'predicted1',
				-source=>'predicted');
my $predicted_exon2 = $ftr->new(-start=>60,-stop=>100,
				-name=>'predicted2',
				-type=>'exon',-source=>'predicted');

my $confirmed_exon3 = $ftr->new(-start=>150,-stop=>190,
				-type=>'exon',-source=>'confirmed',
			       -name=>'abc123');
my $partial_gene = $ftr->new(-segments=>[$confirmed_exon1,$predicted_exon1,$predicted_exon2,$confirmed_exon3],
			     -name => 'partial gene',
			     -type => 'transcript',
			     -source => '(from a big annotation pipeline)'
			    );
my @segments = $partial_gene->segments;
my $score = 10;
foreach (@segments) {
  $_->score($score);
  $score += 10;
}

my $panel = Bio::Graphics::Panel->new(
#				      -grid => [50,100,150,200,250,300,310,320,330],
				      -gridcolor => 'lightcyan',
				      -grid => 1,
				      -segment => $segment,
#				      -offset => 300,
#				      -length  => 1000,
				      -spacing => 15,
				      -width   => 600,
				      -pad_top  => 20,
				      -pad_bottom  => 20,
				      -pad_left => 20,
				      -pad_right=> 20,
#				      -bgcolor => 'teal',
#				      -key_style => 'between',
				      -key_style => 'bottom',
				     );
my @colors = $panel->color_names();

my $t = $panel->add_track(
			  #		  generic => [$abc3,$zed_27],
			  transcript2 => [$abc3,$zed_27],
			  -label => 1,
			  -bump => 1,
			  -key => 'Prophecies',
			  #		  -tkcolor => $colors[rand @colors],
			 );
$t->configure(-bump=>1);
$panel->add_track($segment,
		  -glyph => 'arrow',
		  -label => 'base pairs',
		  -double => 1,
		  -bump => 0,
		  -height => 10,
		  -arrowstyle=>'regular',
		  -linewidth=>1,
#		  -tkcolor => $colors[rand @colors],
		  -tick => 2,
		 );
$panel->unshift_track(generic => [$segment,$zk154_1,$zk154_2,$zk154_3,[$xyz4,$zed_27]],
		      -label     => sub { my $feature = shift; $feature->sub_SeqFeature>0},
		      -bgcolor   => sub { shift->primary_tag eq 'predicted' ? 'olive' : 'red'},
		      -connector => sub { my $feature = shift;
					  my $type = $feature->primary_tag;
					  $type eq 'group'      ? 'dashed'
					    : $type eq 'transcript' ? 'hat'
					      : $type eq 'alignment'  ? 'solid'
						: undef},
		      -all_callbacks => 1,
		      -connector_color => 'black',
		      -height => 10,
		      -bump => 1,
		      -linewidth=>2,
		      #		  -tkcolor => $colors[rand @colors],
		      -key => 'Signs',
		 );

my $track = $panel->add_track('transcript2',
			      -label   => sub { $_[-1]->level == 0 } ,
			      -connector => sub { return shift->type eq 'group' ? 'dashed' : ''},
			      -point  => 0,
			      -orient => 'N',
			      -height => 8,
			      -base => 1,
			      -relative_coords => 1,
			      -tick  => 2,
			      -all_callbacks => 1,
			      -bgcolor => 'red',
			      -key     => 'Dynamically Added');
$track->add_feature($bigone,$zed_27,$abc3);
$track->add_group($predicted_exon1,$predicted_exon2,$confirmed_exon3);

$panel->add_track(
		  [$abc3,$zed_27,$partial_gene],
		  -bgcolor   => sub { shift->source_tag eq 'predicted' ? 'green' : 'blue'},
		  -glyph   => 'transcript',
#		  -glyph   => sub { my $feature = shift; 
#				    return $feature->source_tag eq 'predicted'
#				      ? 'ellipse' : 'transcript'},
		  -label       => sub { shift->sub_SeqFeature > 0 },
#		  -label       => 1,
#		  -description => sub { shift->sub_SeqFeature > 0 },
		  -description => sub {
		    my $feature = shift;
		    return 1   if $feature->primary_tag eq 'transcript';
		    return '*' if $feature->source_tag eq 'predicted';
		    return;
		  },
		  -font2color  => 'red',
		  -bump => +1,
#		  -tkcolor => $colors[rand @colors],
		  -key => 'Portents',
		 );
$panel->add_track(segments => [$segment,$zk154_1,[$zk154_2,$xyz4]],
		  -label     => 1,
		  -bgcolor   => sub { shift->primary_tag eq 'predicted' ? 'green' : 'blue'},
		  -connector => sub { my $primary_tag = shift->primary_tag;
				      $primary_tag eq 'transcript' ? 'hat'
				    : $primary_tag eq 'alignment'  ? 'solid'
				    : undef},
		  -connector_color => 'black',
		  -height => 10,
		  -bump => 1,
#		  -tkcolor => $colors[rand @colors],
		  -key => 'Signals',
		 );
$panel->add_track(generic => [],
		  -key => 'Foobar');

$panel->add_track(graded_segments => $partial_gene,
		  -bgcolor =>'blue',
		  -label   => 1,
		  -key     => 'Scored thing');

$panel->add_track(diamond => [$segment,$zk154_1,$zk154_2,$zk154_3,$xyz4,$zed_27],
		  -bgcolor =>'blue',
		  -label   => 1,
		  -key     => 'pointy thing');

#print $panel->png;

my $gd    = $panel->gd;
my @boxes = $panel->boxes;
my $red   = $panel->translate_color('red');
for my $box (@boxes) {
  my ($feature,@points) = @$box;
#  $gd->rectangle(@points,$red);
}
#$gd->filledRectangle(0,0,20,200,1);
#$gd->filledRectangle(600-20,0,600,200,1);
print $gd->png;

